File: Real-Time-Search-Workshop/javascript/lab4/lab4.js
/**
 * @fileoverview Advanced JSON operations with the node-redis client
 * @maker Joey Whelan
 */

export class Lab4 {
    
    async run(client) {
        console.log('\n*** Lab 4 - Data Loading ***');
        await client.json.set('warehouse:1', '$', {
            "city": "Boston",
            "location": "42.361145, -71.057083",
            "inventory":[{
                "id": 15970,
                "gender": "Men",
                "season":["Fall", "Winter"],
                "description": "Turtle Check Men Navy Blue Shirt",
                "price": 34.95
            },{
                "id": 59263,
                "gender": "Women",
                "season": ["Fall", "Winter", "Spring", "Summer"],
                "description": "Titan Women Silver Watch",
                "price": 129.99
            },{
                "id": 46885,
                "gender": "Boys",
                "season": ["Fall"],
                "description": 
                "Ben 10 Boys Navy Blue Slippers",
                "price": 45.99
        }]});
    
        console.log('\n*** Lab 4 - All Properties of Array ***');
        let result = await client.json.get('warehouse:1', { path: '$.inventory[*]' });
        console.log(JSON.stringify(result, null, 4)); 

        console.log('\n*** Lab 4 - All Properties of a Field ***');
        result = await client.json.get('warehouse:1', { path: '$.inventory[*].price' });
        console.log(JSON.stringify(result, null, 4)); 

        console.log('\n*** Lab 4 - Relational - Equality ***');
        result = await client.json.get('warehouse:1', { path: '$.inventory[?(@.description=="Turtle Check Men Navy Blue Shirt")]' });
        console.log(JSON.stringify(result, null, 4)); 

        console.log('\n*** Lab 4 - Relational - Less Than ***');
        result = await client.json.get('warehouse:1', { path: '$.inventory[?(@.price<100)]' });
        console.log(JSON.stringify(result, null, 4)); 

        console.log('\n*** Lab 4 - Relational - Greater Than or Equal ***');
        result = await client.json.get('warehouse:1', { path: '$.inventory[?(@.id>=20000)]' });
        console.log(JSON.stringify(result, null, 4)); 

        console.log('\n*** Lab 4 - Logical AND ***');
        result = await client.json.get('warehouse:1', { path: '$.inventory[?(@.gender=="Men"&&@.price>20)]' });
        console.log(JSON.stringify(result, null, 4)); 

        console.log('\n*** Lab 4 - Logical OR ***');
        result = await client.json.get('warehouse:1', { path: '$.inventory[?(@.price<100||@.gender=="Women")].id' });
        console.log(JSON.stringify(result, null, 4));

        console.log('\n*** Lab 4 - Regex - Contains Exact ***');
        result = await client.json.get('warehouse:1', { path: '$.inventory[?(@.description =~ "Blue")]' });
        console.log(JSON.stringify(result, null, 4));

        console.log('\n*** Lab 4 - Regex - Contains, Case Insensitive ***');
        result = await client.json.get('warehouse:1', { path: '$.inventory[?(@.description =~ "(?i)watch")]' });
        console.log(JSON.stringify(result, null, 4));

        console.log('\n*** Lab 4 - Regex - Begins With ***');
        result = await client.json.get('warehouse:1', { path: '$.inventory[?(@.description =~ "^T")]' });
        console.log(JSON.stringify(result, null, 4));
    }
}

File: Real-Time-Query-Workshop/python/code/lab4.py
'''
  @fileoverview Advanced JSON operations with the redis client
  @maker Joey Whelan
'''
from redis.commands.json.path import Path
import json

class Lab4(object):
    
    def run(self, client):
        print('\n*** Lab 4 - Data Loading ***')
        client.json().set('warehouse:1', '$', {
            "city": "Boston",
            "location": "42.361145, -71.057083",
            "inventory":[{
                "id": 15970,
                "gender": "Men",
                "season":["Fall", "Winter"],
                "description": "Turtle Check Men Navy Blue Shirt",
                "price": 34.95
            },{
                "id": 59263,
                "gender": "Women",
                "season": ["Fall", "Winter", "Spring", "Summer"],
                "description": "Titan Women Silver Watch",
                "price": 129.99
            },{
                "id": 46885,
                "gender": "Boys",
                "season": ["Fall"],
                "description": 
                "Ben 10 Boys Navy Blue Slippers",
                "price": 45.99
        }]})
    
        print('\n*** Lab 4 - All Properties of Array ***')
        result = client.json().get('warehouse:1', Path('$.inventory[*]'))
        print(json.dumps(result, indent=4)) 

        print('\n*** Lab 4 - All Properties of a Field ***')
        result = client.json().get('warehouse:1', Path('$.inventory[*].price'))
        print(json.dumps(result, indent=4)) 

        print('\n*** Lab 4 - Relational - Equality ***')
        result = client.json().get('warehouse:1', Path('$.inventory[?(@.description=="Turtle Check Men Navy Blue Shirt")]'))
        print(json.dumps(result, indent=4)) 

        print('\n*** Lab 4 - Relational - Less Than ***')
        result = client.json().get('warehouse:1', Path('$.inventory[?(@.price<100)]'))
        print(json.dumps(result, indent=4)) 

        print('\n*** Lab 4 - Relational - Greater Than or Equal ***')
        result = client.json().get('warehouse:1', Path('$.inventory[?(@.id>=20000)]'))
        print(json.dumps(result, indent=4)) 

        print('\n*** Lab 4 - Logical AND ***')
        result = client.json().get('warehouse:1', Path('$.inventory[?(@.gender=="Men"&&@.price>20)]'))
        print(json.dumps(result, indent=4)) 

        print('\n*** Lab 4 - Logical OR ***')
        result = client.json().get('warehouse:1', Path('$.inventory[?(@.price<100||@.gender=="Women")].id'))
        print(json.dumps(result, indent=4))

        print('\n*** Lab 4 - Regex - Contains Exact ***')
        result = client.json().get('warehouse:1', Path('$.inventory[?(@.description =~ "Blue")]'))
        print(json.dumps(result, indent=4))

        print('\n*** Lab 4 - Regex - Contains, Case Insensitive ***')
        result = client.json().get('warehouse:1', Path('$.inventory[?(@.description =~ "(?i)watch")]'))
        print(json.dumps(result, indent=4))

        print('\n*** Lab 4 - Regex - Begins With ***')
        result = client.json().get('warehouse:1', Path('$.inventory[?(@.description =~ "^T")]'))
        print(json.dumps(result, indent=4))